File: R/myRandomForest.R
setwd("~/Documents/GitHub/BikeSharingDemand/R")
train <- read.csv("../csv/train.csv")
test <- read.csv("../csv/test.csv")

library('party')
library('randomForest')

#factorize training set
train_factor <- train
train_factor$weather <- factor(train$weather)
train_factor$holiday <- factor(train$holiday)
train_factor$workingday <- factor(train$workingday)
train_factor$season <- factor(train$season)

#factorize test set
test_factor <- test
test_factor$weather <- factor(test$weather)
test_factor$holiday <- factor(test$holiday)
test_factor$workingday <- factor(test$workingday)
test_factor$season <- factor(test$season)

#create time column by stripping out timestamp
train_factor$time <- substring(train$datetime,12,20)
test_factor$time <- substring(test$datetime,12,20)

#factorize new timestamp column
train_factor$time <- factor(train_factor$time)
test_factor$time <- factor(test_factor$time)

#create day of week column
train_factor$day <- weekdays(as.Date(train_factor$datetime))
train_factor$day <- as.factor(train_factor$day)
test_factor$day <- weekdays(as.Date(test_factor$datetime))
test_factor$day <- as.factor(test_factor$day)

#create Sunday variable
train_factor$sunday[train_factor$day == "Sunday"] <- "1"
train_factor$sunday[train_factor$day != "1"] <- "0"

test_factor$sunday[test_factor$day == "Sunday"] <- "1"
test_factor$sunday[test_factor$day != "1"] <- "0"

#convert to factor
train_factor$sunday <- as.factor(train_factor$sunday)
test_factor$sunday <- as.factor(test_factor$sunday)

#convert time and create $hour as integer to evaluate for daypart
train_factor$hour<- as.numeric(substr(train_factor$time,1,2))
test_factor$hour<- as.numeric(substr(test_factor$time,1,2))

#create daypart column, default to 4 to make things easier for ourselves
train_factor$daypart <- "4"
test_factor$daypart <- "4"


#4AM - 10AM = 1
train_factor$daypart[(train_factor$hour < 10) & (train_factor$hour > 3)] <- 1
test_factor$daypart[(test_factor$hour < 10) & (test_factor$hour > 3)] <- 1


#11AM - 3PM = 2
train_factor$daypart[(train_factor$hour < 16) & (train_factor$hour > 9)] <- 2
test_factor$daypart[(test_factor$hour < 16) & (test_factor$hour > 9)] <- 2


#4PM - 9PM = 3
train_factor$daypart[(train_factor$hour < 22) & (train_factor$hour > 15)] <- 3
test_factor$daypart[(test_factor$hour < 22) & (test_factor$hour > 15)] <- 3

#convert daypart to factor
train_factor$daypart <- as.factor(train_factor$daypart)
test_factor$daypart <- as.factor(test_factor$daypart)

#convert hour back to factor
train_factor$hour <- as.factor(train_factor$hour)
test_factor$hour <- as.factor(test_factor$hour)



#create our formula
formula <- count ~ season + holiday + workingday + weather + temp + atemp + humidity + hour + daypart + sunday

#build our model
# Build Random Forest Ensemble
set.seed(415)
fit <- randomForest(formula, data=train_factor, importance=TRUE, ntree=500)

# Look at variable importance
varImpPlot(fit)
# Now let's make a prediction and write a submission file
Prediction <- predict(fit, test_factor)
submit <- data.frame(datetime = test$datetime, count = Prediction)
write.csv(submit, file = "firstforest.csv", row.names = FALSE)

File: python/myRandomForest.py
import pandas as pd
import numpy as np
import csv as csv
from sklearn.ensemble import RandomForestClassifier

# Data cleanup
train_df = pd.read_csv('../csv/train.csv', header=0)        # Load the train file into a dataframe
#test_df = pd.read_csv('../csv/test.csv', header=0)        # Load the train file into a dataframe

def formatTemp(temp):
    if temp < 10:
        return 1
    elif temp >= 10 and temp < 20:
        return 2
    elif temp >= 20 and temp < 30:
        return 3
    elif temp >= 30:
        return 4

def formatHumidity(humid):
    if humid < 25:
        return 1
    elif humid >= 25 and humid < 50:
        return 2
    elif humid >= 50 and humid < 75:
        return 3
    elif humid >= 75:
        return 4

def formatWindSpeed(ws):
    if ws < 7:
        return 1
    elif ws >= 7 and ws < 10:
        return 2
    elif ws >= 10 and ws < 13:
        return 3
    elif ws >= 13 and ws < 16:
        return 4
    elif ws >= 16 and ws < 19:
        return 5
    elif ws >= 19 and ws < 30:
        return 6
    elif ws >= 30:
        return 7


train_df['temp2'] = train_df['temp'].map(formatTemp).astype(int)
train_df['atemp2'] = train_df['atemp'].map(formatTemp).astype(int)
train_df['humidity2'] = train_df['humidity'].map(formatHumidity).astype(int)
train_df['windspeed2'] = train_df['windspeed'].map(formatWindSpeed).astype(int)
train_df['datetime2'] = pd.to_datetime(train_df['datetime'], dayfirst=True)

print train_df['datetime2'][1]
temp =  type(train_df["datetime2"][1].time())
print temp.hour()
